File: Source/Aladdin.ASN1.ISO/ASN1/ISO/PKCS/PKCS07/RecipientInfos.cs
using System;

// RecipientInfos ::= SET SIZE OF RecipientInfo

namespace Aladdin.ASN1.ISO.PKCS.PKCS7
{
	public class RecipientInfos : Set
	{
		// конструктор при раскодировании
		public RecipientInfos(IEncodable encodable) : 
			base(new ChoiceCreator<RecipientInfo>().Factory(), encodable) {}

		// конструктор при закодировании
		public RecipientInfos(params IEncodable[] values) : 
			base(new ChoiceCreator<RecipientInfo>().Factory(), values) {}

	    // найти информацию отдельного пользователя
	    public IEncodable this[PKIX.Certificate recipientCertificate] { get 
        {
            // получить содержимое сертификата
            PKIX.TBSCertificate tbsCertificate = recipientCertificate.TBSCertificate; 
        
		    // получить идентификатор ключа
		    OctetString keyID = (tbsCertificate.Extensions != null) ? 
                tbsCertificate.Extensions.SubjectKeyIdentifier : null; 
        
            // найти информацию о ключе по идентификатору
		    if (keyID != null) { IEncodable recipientInfo = this[keyID];
        
                // вернуть найденную информацию
                if (recipientInfo != null) return recipientInfo; 
            }
            // найти информацию о ключе по идентификатору
            return this[new PKIX.IssuerSerialNumber(
                tbsCertificate.Issuer, tbsCertificate.SerialNumber
            )];
        }}
		// найти информацию отдельного пользователя
		public IEncodable this[PKIX.IssuerSerialNumber recipientID] { get
		{
            // для всех получателей
			foreach (IEncodable recipientInfo in this)
			{
				// в зависимости от типа
				if (recipientInfo.Tag == Tag.Sequence)
				{
					// преобразовать тип
					KeyTransRecipientInfo info = new KeyTransRecipientInfo(recipientInfo); 

					// проверить совпадение типа
					if (info.Rid.Tag != Tag.Sequence) continue; 

					// проверить совпадение пользователей
					if (recipientID.Equals(info.Rid)) return recipientInfo; 
				}
				// в зависимости от типа
				else if (recipientInfo.Tag == Tag.Context(1))
				{
					// преобразовать тип
					KeyAgreeRecipientInfo info = new KeyAgreeRecipientInfo(recipientInfo); 

              	    // проверить совпадение пользователей
                    if (info.RecipientEncryptedKeys[recipientID] != null) return recipientInfo; 
				}
			}
			return null; 
		}}
		// найти информацию отдельного пользователя
		public IEncodable this[OctetString keyID] { get 
		{
            // для всех получателей
			foreach (IEncodable recipientInfo in this)
			{
				// в зависимости от типа
				if (recipientInfo.Tag == Tag.Sequence)
				{
					// преобразовать тип
					KeyTransRecipientInfo info = new KeyTransRecipientInfo(recipientInfo); 

					// проверить совпадение типа
					if (info.Rid.Tag != Tag.Context(0)) continue; 

					// проверить совпадение пользователей
					if (Arrays.Equals(keyID.Value, info.Rid.Content)) return recipientInfo; 
				}
				// в зависимости от типа
				else if (recipientInfo.Tag == Tag.Context(1))
				{
					// преобразовать тип
					KeyAgreeRecipientInfo info = new KeyAgreeRecipientInfo(recipientInfo); 

              	    // проверить совпадение пользователей
                    if (info.RecipientEncryptedKeys[keyID] != null) return recipientInfo; 
				}
				// в зависимости от типа
				else if (recipientInfo.Tag == Tag.Context(2))
				{
					// преобразовать тип
					KEKRecipientInfo info = new KEKRecipientInfo(recipientInfo); 

					// проверить совпадение пользователей
					if (keyID.Equals(info.Kekid.KeyIdentifier)) return recipientInfo; 
				}
			}
			return null; 
		}}
	}
}

File: Source/Aladdin.ASN1.ISO.Java/src/aladdin/asn1/iso/pkcs/pkcs7/RecipientInfos.java
package aladdin.asn1.iso.pkcs.pkcs7;
import aladdin.asn1.*; 
import aladdin.asn1.Set; 
import aladdin.asn1.iso.pkix.*; 
import java.io.*; 
import java.util.*; 

// RecipientInfos ::= SET SIZE OF RecipientInfo

public final class RecipientInfos extends Set<IEncodable>
{
    private static final long serialVersionUID = 8264260148302910480L;
    
	// конструктор при раскодировании
	public RecipientInfos(IEncodable encodable) throws IOException 
	{
		super(new ChoiceCreator(RecipientInfo.class).factory(), encodable); 
	}
	// конструктор при закодировании
	public RecipientInfos(IEncodable... values) 
	{
		super(new ChoiceCreator(RecipientInfo.class).factory(), values); 
	}
	// найти информацию отдельного пользователя
	public final IEncodable get(Certificate recipientCertificate) throws IOException
    {
        // получить содержимое сертификата
        TBSCertificate tbsCertificate = recipientCertificate.tbsCertificate(); 
        
		// получить идентификатор ключа
		OctetString keyID = (tbsCertificate.extensions() != null) ?
            tbsCertificate.extensions().subjectKeyIdentifier() : null; 
        
        // найти информацию о ключе по идентификатору
		if (keyID != null) { IEncodable recipientInfo = get(keyID);
        
            // вернуть найденную информацию
            if (recipientInfo != null) return recipientInfo; 
        }
        // найти информацию о ключе по идентификатору
        return get(new IssuerSerialNumber(
            tbsCertificate.issuer(), tbsCertificate.serialNumber()
        ));
    }
	// найти информацию отдельного пользователя
	public final IEncodable get(IssuerSerialNumber recipientID) throws IOException
	{
        // для всех получателей
		for (IEncodable recipientInfo : this)
		{
			// в зависимости от типа
			if (recipientInfo.tag().equals(Tag.SEQUENCE))
			{
				// преобразовать тип
				KeyTransRecipientInfo info = new KeyTransRecipientInfo(recipientInfo); 

				// проверить совпадение типа
				if (!info.rid().tag().equals(Tag.SEQUENCE)) continue; 

				// проверить совпадение пользователей
				if (recipientID.equals(info.rid())) return recipientInfo; 
			}
			// в зависимости от типа
			else if (recipientInfo.tag().equals(Tag.context(1)))
			{
				// преобразовать тип
				KeyAgreeRecipientInfo info = new KeyAgreeRecipientInfo(recipientInfo); 
                
              	// проверить совпадение пользователей
                if (info.recipientEncryptedKeys().get(recipientID) != null) return recipientInfo; 
			}
		}
		return null; 
	}
	// найти информацию отдельного пользователя
	public final IEncodable get(OctetString keyID) throws IOException
	{
        // для всех получателей
		for (IEncodable recipientInfo : this)
		{
			// в зависимости от типа
			if (recipientInfo.tag().equals(Tag.SEQUENCE))
			{
				// преобразовать тип
				KeyTransRecipientInfo info = new KeyTransRecipientInfo(recipientInfo); 

				// проверить совпадение типа
				if (!info.rid().tag().equals(Tag.context(0))) continue; 
                
				// проверить совпадение пользователей
				if (Arrays.equals(keyID.value(), info.rid().content())) return recipientInfo; 
			}
			// в зависимости от типа
			else if (recipientInfo.tag().equals(Tag.context(1)))
			{
				// преобразовать тип
				KeyAgreeRecipientInfo info = new KeyAgreeRecipientInfo(recipientInfo); 

              	// проверить совпадение пользователей
                if (info.recipientEncryptedKeys().get(keyID) != null) return recipientInfo; 
			}
			// в зависимости от типа
			else if (recipientInfo.tag().equals(Tag.context(2)))
			{
				// преобразовать тип
				KEKRecipientInfo info = new KEKRecipientInfo(recipientInfo); 

				// проверить совпадение пользователей
				if (keyID.equals(info.kekId().keyIdentifier())) return recipientInfo; 
			}
		}
		return null; 
	}
}
